### PR TITLE
Close changesets in background

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes #13367 by moving the closing of changesets out of the request path (when applying a campaign spec or closing a campaign) and into the background, into the reconciler.

It does so by introducing yet another boolean flag on the `changesets` table called `closing`, which is set to `true` along with `reconciler_state` to `'queued'` when the changeset should be closed.

Just like in #13288 I couldn't see a way to avoid that boolean that doesn't introduce another edge case somewhere else.

One option would be, for example, to add a `closed: Boolean` to the `ChangesetSpec` schema (which we don't have to expose to the users, but which could allow users to close single changesets) and then set that. Problem is: our `ChangesetSpecs` are immutable right now. When you want to change one, you need a new one and apply that. So if I want to keep that immutability (which seems like a nice property to preserve) I'd have to create a new `ChangesetSpec` when I want to close the changeset and (here's the kicker) a new `CampaignSpec` to which to attach it... Then I'd have to somehow apply that along with all the other changeset specs, etc.

I still think there might be a way to make this idea of "we now want to change the state, internally. reconciler, make it happen" flow better through the system, where "unsynced" and "closing" might be combined, but I haven't found it yet. If you have better ideas: let me know!

### CODE REVIEW

File: enterprise/internal/campaigns/reconciler.go [hunk 1]
```diff
@@ -240,6 +244,34 @@ func (r *reconciler) updateChangeset(ctx context.Context, tx *Store, ch *campaig
...
+	}
+
+	cs := &repos.Changeset{Changeset: ch}
+
+	if err := ccs.CloseChangeset(ctx, cs); err != nil {
+		return errors.Wrap(err, "creating changeset")
+	}
+
+	fmt.Printf("we just closed the changeset! cs.State=%s\n", cs.Changeset.Metadata.(*github.PullRequest).State)
+
```
Comment (eseliger) [inline enterprise/internal/campaigns/reconciler.go:267]: do we want to retain that?

File: enterprise/internal/campaigns/service_apply_campaign.go [hunk 2]
```diff
@@ -57,35 +50,6 @@ func (s *Service) ApplyCampaign(ctx context.Context, opts ApplyCampaignOpts) (ca
...
-			ctx := trace.ContextWithTrace(context.Background(), tr)
-
-			// Close only the changesets that are open
-			err := s.CloseOpenChangesets(ctx, toClose)
-			if err != nil {
-				log15.Error("CloseCampaignChangesets", "err", err)
-			}
-		}()
-	}()
-
```
Comment (eseliger) [inline enterprise/internal/campaigns/service_apply_campaign.go:88]: So happy to see this disappear!

**Review by eseliger — APPROVED** [review]
Nice :) That was easier than expected

File: enterprise/internal/campaigns/service_apply_campaign.go (hunk 2 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/service_apply_campaign.go:88]: Same!

File: enterprise/internal/campaigns/reconciler.go (hunk 1 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/reconciler.go:267]: That's the most polite way I've ever heard someone say "dude, get rid of your debug stuff you left lying around"

**Pull request merged**